### ISSUE TITLE
Release the GIL during read/write

### ISSUE THREAD

**Issue by clbarnes** [pattonw/rust-pyn5#55]
https://github.com/PyO3/pyo3/blob/master/guide/src/parallelism.md

If I understand the GIL correctly, this should lead to a performance improvement when used in a (python) threaded context.

**Comment by clbarnes** [comment]
The trivial approach of adding `py.allow_threads(move || whatever_operation())` doesn't change the runtime of a simple thread-parallelised write op, so I guess the GIL isn't a major factor here.

**Comment by aschampion** [comment]
Still seems really slow. How is the bench being invoked?

**Comment by clbarnes** [comment]
Agreed! The "proper" benchmark uses pytest-benchmark, but with a serial execution time of 2min45 I didn't really want it running hundreds of times...

For running once, I am just using a script:

```python
from concurrent.futures import as_completed
from concurrent.futures.thread import ThreadPoolExecutor
from pathlib import Path
import math
import itertools
import shutil

import numpy as np
import tifffile
from tqdm import tqdm

import pyn5

fpath = Path("tests/data/JeffT1_le.tif")
tiffdata = tifffile.imread(str(fpath))
t1_data = np.asarray(tiffdata, dtype="int16")

block_len = 64
block_reps = 5

desired_len = block_len * block_reps

repeats = tuple(math.ceil(desired_len / d) for d in t1_data.shape)
d = t1_data.astype(np.int16)
for axis, reps in enumerate(repeats):
    d = np.concatenate([d] * reps, axis)

arr = d[:desired_len, :desired_len, :desired_len]

chunks = (block_len,) *3

tmp_path = Path('tmp').resolve()
tmp_path.mkdir(exist_ok=True)

root_path = str(tmp_path / "bench.n5")
path_name = "ds"


def create():
    pyn5.create_dataset(
        root_path,
        path_name,
        list(arr.shape),
        list(chunks),
        arr.dtype.name.upper()
    )

    return pyn5.DatasetINT16(root_path, path_name, False)


def clean():
    shutil.rmtree(root_path, ignore_errors=True)


def write_whole(ds):
    ds.write_ndarray([0,0,0], arr, 0)


def write_block(ds, arr, block_idx):
    slicing = tuple(slice(d * block_len, (d + 1) * block_len) for d in block_idx)
    data = arr[slicing].flatten().tolist()
    ds.write_block(block_idx, data)


def write_by_block(ds):
    for zyx in tqdm(itertools.product(range(block_reps), repeat=3), total=block_reps**3):
        write_block(ds, arr, zyx)


def write_by_block_parallel(ds):
    with ThreadPoolExecutor() as exe:
        futs = []
        for zyx in tqdm(itertools.product(range(block_reps), repeat=3), total=block_reps ** 3, desc="submitting"):
            futs.append(exe.submit(write_block, ds, arr, zyx))

        for _ in tqdm(as_completed(futs), total=len(futs), desc="processing"):
            pass


clean()
ds = create()
```

**Comment by clbarnes** [comment]
z5 does it in serial in 3s; we're only 50x off...

**Comment by aschampion** [comment]
Is z5 using the same gzip level?

**Comment by aschampion** [comment]
For reference:

```sh
test simple_rw_i16_gzip_1 ... bench:   9,052,302 ns/iter (+/- 614,825) = 57 MB/s
test simple_rw_i16_gzip_6 ... bench:  23,030,927 ns/iter (+/- 1,272,153) = 22 MB/s
test simple_rw_i16_gzip_9 ... bench:  23,059,114 ns/iter (+/- 2,027,909) = 22 MB/s
test simple_rw_i16_raw    ... bench:     570,984 ns/iter (+/- 102,884) = 918 MB/s
```

(6 is the default gzip level.)

**Comment by clbarnes** [comment]
I was using the default; I've just checked and [it's level 5](https://github.com/constantinpape/z5/blob/master/src/python/module/z5py/dataset.py#L71).

**Comment by aschampion** [comment]
As always seems to be the case with python, this looks like it may be down to some sort of environment/path magic not doing what you'd expect, because of how pyo3 or our `setup.py` work.

When I run your script from the `rust-pyn5` folder, I get the same awful performance.

```sh
In [2]: %time write_by_block(ds)
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 125/125 [02:08<00:00,  1.02s/it]
CPU times: user 2min 8s, sys: 326 ms, total: 2min 8s
Wall time: 2min 8s
```

When I run it from another folder in a fresh pipenv shell, it takes 5 seconds to serially write all the blocks, similar to z5.

```sh
In [2]: %time write_by_block(ds)
100%|███████████████████████████████████████████████████████████████████████████████| 125/125 [00:05<00:00, 23.72it/s]
CPU times: user 5.16 s, sys: 111 ms, total: 5.27 s
Wall time: 5.28 s
```

I only got clued down this path because I noticed debugging timing code I was putting in `lib.rs` wasn't getting printed, even though `python setup.py install` was obviously compiling and copying the update module.

When run from within the `rust-pyn5` folder, it's using whatever library file is in `pyn5/` from the last time you ran `python setup.py develop` (which is a slow debug build), even if you've run `python setup.py install` since.

**Comment by clbarnes** [comment]
Interesting! I can replicate that (takes 7s on mine, I think my desktop's running off a slow HDD). 

Using `pytest-benchmark` (and therefore `setup.py test`) runs a production build (which is a `setuptools_rust` feature) so I'd assumed it then actually *used* that production build... This may be worth raising an issue on `setuptools_rust`; as you say it's very unexpected behaviour.

The parallel (python threaded) version now gives no speedup over the serial version so I'll have another look at the GIL stuff.

Thanks for looking into this, good to know I'm not going crazy.

**Comment by clbarnes** [comment]
Adding those `allow_threads` closures gets a 3x speedup using a python `ThreadPoolExecutor`, pretty painless.

**Comment by clbarnes** [comment]
Merged in #61
